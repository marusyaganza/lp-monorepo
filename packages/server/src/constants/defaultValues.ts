import { Language, Role, SortWordsBy, Tense } from '../generated/graphql';

export const WORDS_PER_PAGE = 5;
export const DEFAULT_LANGUAGE = Language.English;
export const DEFAULT_WORDS_SORT_BY = SortWordsBy.UpdatedAt;
export const DEFAULT_ROLE = Role.Member;
export const DEFAULT_TENSE = Tense.Pind;
export const DEFAULT_TOKEN_TTL = 7;
// to clean db every hour + 10 min just in case
export const DEMO_DB_TTL = 4200;
export const DEFAULT_MONGO_CONNECTION = 'mongodb://localhost:27017';
export const DEFAULT_OPTIONS__SAMPLE_SIZE = 50;
