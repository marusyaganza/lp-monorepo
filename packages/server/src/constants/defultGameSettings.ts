import { Game, Language, SettingsByGame } from '../generated/graphql';

export const DEFAULT_GAMES_SETTINGS: SettingsByGame = {
  [Game.Audio]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Image]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Speaking]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.TypeWord]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.SelectDef]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 6,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.SelectWord]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 6,
    timesToLearn: 5,
    languages: [Language.English, Language.Spanish]
  },
  [Game.Conjugation]: {
    wordsPerGame: 2,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.Spanish]
  },
  [Game.Gender]: {
    wordsPerGame: 6,
    minWords: 1,
    optionsPerGame: 0,
    timesToLearn: 5,
    languages: [Language.Spanish]
  }
};
