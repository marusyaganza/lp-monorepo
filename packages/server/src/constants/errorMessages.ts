export const ERROR_MESSAGES = {
  NOT_AUTHENTICATED: 'Login to perform this operation',
  NOT_AUTHORIZED: 'You do not have permission to perform this operation',
  INCORRECT_CREDS: 'email or password is incorrect',
  SIGN_UP_FAILED: 'sign up operation failed',
  SING_UP_UNAVAILABLE:
    'Sign up operation is not available now. Sorry for inconvenience.',
  GAME_NOT_FOUND: 'Game not found',
  GAME_GENERATION_FAILED: 'Game data generation failed'
};
