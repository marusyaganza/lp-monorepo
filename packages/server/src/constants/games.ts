import { Game, GameConfig, Language } from '../generated/graphql';
export const GAMES: GameConfig[] = [
  {
    name: 'WordEcho',
    id: 'audio',
    desc: 'Listen, Type, Master. Sharpen language skills through auditory engagement.',
    type: Game.Audio,
    orderNum: 0,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'LexiSense',
    desc: 'Define & Conquer.Match words with their meanings intuitively.',
    id: 'select_definition',
    type: Game.SelectDef,
    orderNum: 1,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'Definify: Match & Learn',
    desc: 'Associate definitions with the right words. A captivating word puzzle adventure!',
    id: 'select_word',
    type: Game.SelectWord,
    orderNum: 2,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'WordRecall: Define & Type',
    desc: 'Type words matching given definitions. Boost vocabulary in engaging rounds!',
    id: 'type_word',
    type: Game.TypeWord,
    orderNum: 3,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'PictureWord Blitz',
    desc: 'Identify objects from pictures and type their names to enhance vocabulary skills!',
    id: 'image',
    type: Game.Image,
    orderNum: 4,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'VoiceWord Challenge',
    desc: 'Speak words from clues and images. Sharpen your vocabulary and speaking skills fast!',
    id: 'speaking',
    type: Game.Speaking,
    orderNum: 5,
    languages: [Language.English, Language.Spanish]
  },
  {
    name: 'Conjugation Quest',
    desc: 'Master verb conjugations across tenses and languages in fast-paced, interactive challenges!',
    id: 'conjugation',
    type: Game.Conjugation,
    orderNum: 6,
    languages: [Language.Spanish]
  },
  {
    name: 'GenderWord Challenge',
    desc: 'Choose nouns and match their gender. Test linguistic skills in thrilling rounds!',
    id: 'gender',
    type: Game.Gender,
    orderNum: 7,
    languages: [Language.Spanish]
  }
];
