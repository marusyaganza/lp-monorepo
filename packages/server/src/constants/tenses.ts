import { Tense } from '../generated/graphql';

// Uncomment more tenses as needed
export const TENSES: Record<Tense, string> = {
  pind: 'Presente indicativo',
  pprf: 'Pretérito perfecto simple',
  impf: 'Imperativo',
  pret: 'Pretérito imperfecto',
  futr: 'Futuro indicativo',
  cond: 'Condicional',
  psub: 'Presente subjentivo'
  // gppt: 'Gerundio'
  //   pisb1: 'Pretérito imperfecto',
  //   pisb2: 'Pretérito imperfecto 2',
  //   fsub: 'Futuro',
  //   ppci: 'Pretérito perfecto compuesto',
  //   ppsi: 'Pretérito pluscuamperfecto',
  //   pant: 'Pretérito anterior',
  //   fpin: 'Futuro perfecto',
  //   cpef: 'Condicional perfecto',
  //   ppfs: 'Pretérito perfecto',
  //   ppss1: 'Pretérito pluscuamperfecto',
  //   ppss2: 'Pretérito pluscuamperfecto 2',
  //   fpsb: 'Futuro perfecto',
};
