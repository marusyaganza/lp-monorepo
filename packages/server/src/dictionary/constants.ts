import { Tense } from 'generated/graphql';

export type TagType = {
  tag: string;
  replacement: string;
};

export type ComplexTagType = {
  opening: string;
  closing: string;
};

export const TAGS: TagType[] = [
  { tag: '{b}', replacement: '<b>' },
  { tag: '{/b}', replacement: '</b>' },
  { tag: '{ldquo}', replacement: '\u201C' },
  { tag: '{rdquo}', replacement: '\u201D' },
  { tag: '{it}', replacement: '<i>' },
  { tag: '{/it}', replacement: '</i>' },
  { tag: '{wi}', replacement: '<i>' },
  { tag: '{/wi}', replacement: '</i>' },
  { tag: '{gloss}=', replacement: '<i>' },
  { tag: '{/gloss}', replacement: '</i>' },
  { tag: '{sc}', replacement: '<i>' },
  { tag: '{/sc}', replacement: '</i>' },
  { tag: '{parahw}', replacement: '</i>' },
  { tag: '{/parahw}', replacement: '</i>' },
  { tag: '{phrase}', replacement: '</i>' },
  { tag: '{/phrase}', replacement: '</i>' },
  { tag: '{qword}', replacement: '</i>' },
  { tag: '{/qword}', replacement: '</i>' },
  { tag: '{itsc}', replacement: '<i>' },
  { tag: '{/itsc}', replacement: '</i>' },
  { tag: '{inf}', replacement: '' },
  { tag: '{/inf}', replacement: '' },
  { tag: '{sup}', replacement: '' },
  { tag: '{/sup}', replacement: '' },
  { tag: '{bc}', replacement: '' },
  { tag: '{dx}', replacement: '' },
  { tag: '{/dx}', replacement: '' },
  { tag: '{dx_def}', replacement: '' },
  { tag: '{/dx_def}', replacement: '' },
  { tag: '{dx_ety}', replacement: '' },
  { tag: '{/dx_ety}', replacement: '' },
  { tag: '{ma}', replacement: '' },
  { tag: '{/ma}', replacement: '' },
  { tag: '{rom}', replacement: '' },
  { tag: '{/rom}', replacement: '' }
];

export const COMPLEX_TAGS: ComplexTagType[] = [
  { opening: '{sx', closing: '}' },
  { opening: '{d_link', closing: '}' },
  { opening: '{a_link', closing: '}' },
  { opening: '{i_link', closing: '}' },
  { opening: '{et_link', closing: '}' },
  { opening: '{mat', closing: '}' },
  { opening: '{dxt', closing: '}' }
];

export const COMPLEX_TAGS_TO_REMOVE: ComplexTagType[] = [
  { opening: '{ds|', closing: '}' }
];

export const HW_TAGS: TagType[] = [
  { tag: '{bit}', replacement: '' },
  { tag: '{/bit}', replacement: '' }
];

// Uncomment more tenses as needed
export const TENSES: Record<Tense, string> = {
  pind: 'Presente indicativo',
  pprf: 'Pretérito perfecto simple',
  impf: 'Imperativo',
  pret: 'Pretérito imperfecto'
  //   futr: 'Futuro indicativo',
  //   cond: 'Condicional',
  //   psub: 'Presente subjentivo',
  //   pisb1: 'Pretérito imperfecto',
  //   pisb2: 'Pretérito imperfecto 2',
  //   fsub: 'Futuro',
  //   ppci: 'Pretérito perfecto compuesto',
  //   ppsi: 'Pretérito pluscuamperfecto',
  //   pant: 'Pretérito anterior',
  //   fpin: 'Futuro perfecto',
  //   cpef: 'Condicional perfecto',
  //   ppfs: 'Pretérito perfecto',
  //   ppss1: 'Pretérito pluscuamperfecto',
  //   ppss2: 'Pretérito pluscuamperfecto 2',
  //   fpsb: 'Futuro perfecto',
};
