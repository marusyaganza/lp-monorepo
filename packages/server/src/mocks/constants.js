export const mockEnv = {
  SALT_NUMBER: 3,
  JWT_SECTET: 'JWT_secret'
};
