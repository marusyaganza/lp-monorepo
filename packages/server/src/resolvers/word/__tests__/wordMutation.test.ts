import {
  createTestServer,
  getDataFromGQL,
  getErrorMessageFromGQL
} from '../../../tests/helpers';
import { mutations } from '../../../tests/mocks/gqlMutations';
import { models } from '../../../tests/mocks/mockModels';
import { testData } from '../../../tests/mocks/dbTestData';
import { ERROR_MESSAGES } from '../../../constants/errorMessages';
import { Game } from '../../../generated/graphql';

const user = { id: 'userId' };

const gameResultData = [
  {
    id: '1',
    hasError: true,
    gameType: Game.Audio
  },
  {
    id: '2',
    hasError: false,
    gameType: Game.Audio
  },
  {
    id: '3',
    hasError: false,
    gameType: Game.Audio
  }
];

describe('word mutations', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  afterAll(() => {
    jest.resetAllMocks();
  });

  test('delete word', async () => {
    const deleteOne = jest.fn(() => ({ ok: true }));
    const { mutate } = createTestServer({
      user,
      models: {
        ...models,
        Word: {
          deleteOne
        }
      }
    });

    const res = await mutate({
      mutation: mutations.deleteWordMutation,
      variables: {
        deleteWordId: 'deleteWordId'
      }
    });
    expect(getDataFromGQL(res).deleteWord).toEqual(
      'word with id deleteWordId was deleted'
    );
    expect(getErrorMessageFromGQL(res)).toEqual(undefined);
    expect(deleteOne).toHaveBeenCalledWith('deleteWordId', 'userId');
  });

  test('delete word if operation fails', async () => {
    const deleteOne = jest.fn(() => ({ ok: false }));
    const { mutate } = createTestServer({
      user,
      models: {
        ...models,
        Word: {
          deleteOne
        }
      }
    });

    const res = await mutate({
      mutation: mutations.deleteWordMutation,
      variables: {
        deleteWordId: 'deleteWordId'
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(
      'deleting word with id deleteWordId failed'
    );
    expect(deleteOne).toHaveBeenCalledWith('deleteWordId', 'userId');
  });

  test('delete word if operation result is undefined', async () => {
    const { mutate } = createTestServer({
      user,
      models
    });

    const res = await mutate({
      mutation: mutations.deleteWordMutation,
      variables: {
        deleteWordId: 'deleteWordId'
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(
      'deleting word with id deleteWordId failed'
    );
    expect(models.Word.deleteOne).toHaveBeenCalledWith(
      'deleteWordId',
      'userId'
    );
  });

  test('delete word if user is not defined', async () => {
    const deleteOne = jest.fn(() => ({ ok: true }));
    const { mutate } = createTestServer({
      models: {
        ...models,
        Word: {
          deleteOne
        }
      }
    });

    const res = await mutate({
      mutation: mutations.deleteWordMutation,
      variables: {
        deleteWordId: 'deleteWordId'
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(ERROR_MESSAGES.NOT_AUTHORIZED);
    expect(deleteOne).toHaveBeenCalledTimes(0);
  });

  test('create word', async () => {
    const { mutate } = createTestServer({
      user,
      models
    });

    const res = await mutate({
      mutation: mutations.saveWordMutation,
      variables: {
        input: {
          ...testData.createWordInput
        }
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(undefined);
    expect(models.Word.createOne).toHaveBeenCalledWith(
      testData.createWordInput,
      'userId'
    );
  });

  test('not create word if user is not defined', async () => {
    const { mutate } = createTestServer({
      models
    });

    const res = await mutate({
      mutation: mutations.saveWordMutation,
      variables: {
        input: {
          ...testData.createWordInput
        }
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(ERROR_MESSAGES.NOT_AUTHORIZED);
    expect(models.User.findOne).toHaveBeenCalledTimes(0);
    expect(models.Word.createOne).toHaveBeenCalledTimes(0);
  });

  test('create word if operation returns undefined', async () => {
    // @ts-expect-error: ts do not hanle this mocking case
    models.Word.createOne.mockReturnValueOnce(undefined);
    const { mutate } = createTestServer({
      user,
      models
    });

    const res = await mutate({
      mutation: mutations.saveWordMutation,
      variables: {
        input: {
          ...testData.createWordInput
        }
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(
      'saving word with name egalitarian failed'
    );
    expect(models.Word.createOne).toHaveBeenCalledTimes(1);
    expect(models.Word.createOne).toHaveBeenCalledWith(
      testData.createWordInput,
      'userId'
    );
  });

  test('update word', async () => {
    const updateOne = jest.fn(() => ({
      ok: true,
      value: { ...testData.createWordInput2, user: 'mockUserId' }
    }));
    const { mutate } = createTestServer({
      user,
      models: {
        ...models,
        Word: {
          updateOne
        }
      }
    });

    const res = await mutate({
      mutation: mutations.updateWordMutation,
      variables: {
        input: { ...testData.updateWordInput, id: '1' }
      }
    });
    expect(getDataFromGQL(res)).toMatchSnapshot();
    expect(getErrorMessageFromGQL(res)).toEqual(undefined);
    expect(updateOne).toHaveBeenCalledWith(
      {
        ...testData.updateWordInput,
        id: '1'
      },
      'userId'
    );
  });

  test('update word if operation fails', async () => {
    const updateOne = jest.fn(() => ({ ok: false }));
    const { mutate } = createTestServer({
      user,
      models: {
        ...models,
        Word: {
          updateOne
        }
      }
    });

    const res = await mutate({
      mutation: mutations.updateWordMutation,
      variables: {
        input: { ...testData.updateWordInput, id: '1' }
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(
      'updating word with id 1 failed'
    );
    expect(updateOne).toHaveBeenCalledWith(
      {
        ...testData.updateWordInput,
        id: '1'
      },
      'userId'
    );
  });

  test('update word if operation result is undefined', async () => {
    const updateOne = jest.fn();
    const { mutate } = createTestServer({
      user,
      models: {
        ...models,
        Word: {
          updateOne
        }
      }
    });

    const res = await mutate({
      mutation: mutations.updateWordMutation,
      variables: {
        input: { ...testData.updateWordInput, id: '1' }
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(
      'updating word with id 1 failed'
    );
    expect(updateOne).toHaveBeenCalledWith(
      {
        ...testData.updateWordInput,
        id: '1'
      },
      'userId'
    );
  });

  test('update word if user is not defined', async () => {
    const updateOne = jest.fn(() => ({ ok: true }));
    const { mutate } = createTestServer({
      models: {
        ...models,
        Word: {
          updateOne
        }
      }
    });

    const res = await mutate({
      mutation: mutations.updateWordMutation,
      variables: {
        input: { ...testData.updateWordInput, id: '1' }
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(ERROR_MESSAGES.NOT_AUTHORIZED);
    expect(updateOne).toHaveBeenCalledTimes(0);
  });
  test('safe game result', async () => {
    const updateStatistics = jest.fn(() => ({ ok: true }));
    const { mutate } = createTestServer({
      user,
      models: {
        ...models,
        Word: {
          updateStatistics
        }
      }
    });

    const res = await mutate({
      mutation: mutations.saveGameResultMutation,
      variables: {
        input: gameResultData
      }
    });
    expect(getDataFromGQL(res).saveGameResult).toEqual('Game result saved');
    expect(getErrorMessageFromGQL(res)).toEqual(undefined);
    expect(updateStatistics).toHaveBeenCalledWith(gameResultData, user.id);
  });

  test('safe game result if operation failed', async () => {
    const updateStatistics = jest.fn(() => ({ ok: false }));
    const { mutate } = createTestServer({
      user,
      models: {
        ...models,
        Word: {
          updateStatistics
        }
      }
    });

    const res = await mutate({
      mutation: mutations.saveGameResultMutation,
      variables: {
        input: gameResultData
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(
      'saving training result failed'
    );
    expect(updateStatistics).toHaveBeenCalledWith(gameResultData, user.id);
  });

  test('safe game result if operation result is undefined', async () => {
    const updateStatistics = jest.fn();
    const { mutate } = createTestServer({
      user,
      models: {
        ...models,
        Word: {
          updateStatistics
        }
      }
    });

    const res = await mutate({
      mutation: mutations.saveGameResultMutation,
      variables: {
        input: gameResultData
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(
      'saving training result failed'
    );
    expect(updateStatistics).toHaveBeenCalledWith(gameResultData, user.id);
  });

  test('safe game result if user is not defined', async () => {
    const updateStatistics = jest.fn();
    const { mutate } = createTestServer({
      models: {
        ...models,
        Word: {
          updateStatistics
        }
      }
    });

    const res = await mutate({
      mutation: mutations.saveGameResultMutation,
      variables: {
        input: gameResultData
      }
    });
    expect(getErrorMessageFromGQL(res)).toEqual(ERROR_MESSAGES.NOT_AUTHORIZED);
    expect(updateStatistics).toHaveBeenCalledTimes(0);
  });
});
