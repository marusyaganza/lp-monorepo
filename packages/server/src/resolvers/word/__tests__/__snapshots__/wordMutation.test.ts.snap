// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`word mutations update word 1`] = `
{
  "updateWord": {
    "additionalInfo": null,
    "audioUrl": null,
    "defs": [
      {
        "def": "{bc} a strong substance that stretches and that is made out of chemicals or from the juice of a tropical tree",
        "examples": null,
      },
      {
        "def": "{it}British{/it}",
        "examples": null,
      },
      {
        "def": "{b}{it}rubbers{/it}{/b} {it}US{/it}, {it}informal + old-fashioned{/it} {bc} shoes or boots that are made of rubber and that fit over your regular shoes to keep them dry",
        "examples": null,
      },
    ],
    "imgUrl": null,
    "isOffensive": null,
    "level": "B2",
    "name": "rubber",
    "particle": "noun",
    "shortDef": [
      "a thin, flexible loop that is made of rubber and used to hold things together â€”called also (US) elastic, elastic band",
    ],
    "stems": [
      "rubber",
      "rubbers",
      "burn rubber",
    ],
    "tags": null,
    "transcription": null,
    "uuid": "59e03286-73c9-43e6-b1eb-4407a6c06b1e",
  },
}
`;
